data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -231,7 +231,23 @@ public void nonNullWrite(int rowId, String data, ColumnVector output) {
 
     @Override
     public void nonNullWrite(int rowId, ByteBuffer data, ColumnVector output) {
-      ((BytesColumnVector) output).setRef(rowId, data.array(), 0, data.array().length);
+      // When there is a backing heap based byte array, we avoid the overhead of
+      // copying it.
+      if (data.hasArray()) {
+        // Don't assume the we're reading in the entire backing array.
+        // Using slice as any mutations to the data at rowId of output
+        // would also be visible in `data`.
+        ByteBuffer slice = data.slice();
+        ((BytesColumnVector) output).setRef(rowId, slice.array(), 0, slice.array().length);
+      } else {
+        // Consume the remaining contents of the input data
+        byte[] bytes = new byte[data.remaining()];
+        data.get(bytes);
+        // Restores the buffer position
+        // TODO - Is this necessary?
+        data.position(data.position() - bytes.length);
```

Comment:
Could you avoid this by using `slice` too? Maybe factoring the slice out of the `if...else`?

Reply:
I agree. By using either `slice` or `duplicate`, there is no need to restore the state of the incoming buffer.

Reply:
Ok. I will prefer `duplicate` here. I'll still need to use a Suppression, but it will be the correct solution if I understand correctly.

Reply:
I actually didn't have to use the suppression once I realized why I needed to use arrayOffset in addition to hasArray. Thanks guys! Let me know if this solution won't do.